lib/elixir/lib/kernel.ex

```diff
@@ -3029,16 +3029,24 @@ defmodule Kernel do
     in_module? = (__CALLER__.context == nil)
 
     right = case bootstrapped?(Macro) do
-      true  -> Macro.expand(right, __CALLER__)
+      true  -> Macro.prewalk(right, &Macro.expand(&1, __CALLER__))
       false -> right
     end
 
     case right do
       [] when not in_module? ->
         false
 
-      [h | t] ->
-        in_var(in_module?, left, &in_list(&1, h, t))
+      [head | tail] when not in_module? ->
+        in_var(in_module?, left, &in_list(&1, head, tail))
+
+      [_ | _] = list when in_module? ->
+        {[head | tail], {_, ast}} = ensure_evaled(list, {0, []})
+        {vars, values} = :lists.unzip(:lists.reverse(ast))
+        quote do
+          {unquote_splicing(vars)} = {unquote_splicing(values)}
```

Comment:
Let's not generate an empty assignment if the ast is empty. It will show up on Macro.expand and it will be confusing.